[CA-930] Don't delete resources with children

andy7i opened the pull request:
Ticket: https://broadworkbench.atlassian.net/browse/CA-930

This PR updates the existing API for delete resource to check for resources with parents or children.

a few notes:
- 403 if the resource has a parent and the user does not have remove_child access on the parent resource
- instead of a new v2 API, this will update the v1 API. Part of the reason for this is that this uses v2 behavior (adding parent info) and also the v1 API would need to be updated with the same or similar behavior to prevent deleting a resource that has children.
- tests for this are being put in v2 spec, partly because it uses v2 behavior, but mostly because v0, v1 behavior should not have changed anyway and we should be putting new tests into the latest version spec file.

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

andy7i committed:
init - added relevant code, hasChildren, and unit test

andy7i committed:
only allow delete if the user also has remove_child action permissions to the parent resource. (based on doug's mac behavior)

andy7i committed:
only allow delete if the user also has remove_child action permissions to the parent resource. (based on doug's mac behavior)
(pt 2 -- forgot to include this)

andy7i committed:
make V2s of deleteResource

andy7i committed:
ResourceServiceSpec happy case - "delete a child resource that has a parent"

andy7i committed:
Remove V2 -- only update V1 instead.

andy7i committed:
Unit tests in the V2 Spec (may need to move these to the V1 spec...)

andy7i committed:
remove test that should no longer be necessary without the v2

andy7i committed:
swagger api update (if we want it)

andy7i committed:
remove "v2" from unit test descriptions

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:
```diff
@@ -196,15 +196,29 @@ class ResourceService(
   def deleteResource(resource: FullyQualifiedResourceId, samRequestContext: SamRequestContext): Future[Unit] =
     for {
+      // First check if children exist. If so, then throw a 400.
+      _ <- hasChildren(resource, samRequestContext).unsafeToFuture() map {
+        case true => throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Cannot delete a resource with children. Delete the children first then try again."))
+        case false =>
+      }
```
I would have a function called `checkNoChildren` instead of `hasChildren` that lists the children and does this check all in one

dvoet replied:
also an `if` statement is easier to follow than a `case` for boolean (I think)

dvoet commented on src/test/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceServiceSpec.scala:
```diff
@@ -715,6 +715,44 @@ class ResourceServiceSpec extends FlatSpec with Matchers with ScalaFutures with
+    assert(policyDAO.listAccessPolicies(childResource, samRequestContext).unsafeRunSync().isEmpty)
+  }
+
+  it should "return 400 for a parent resource that has any children" in {
```
also have a test that checks for 400 in the route spec. In the test here I mostly care that it fails and does not delete the resource. In the route spec we care that the response code is 400.

andy7i committed:
pr feedback -- `checkNoChildren`

andy7i committed:
test description

andy7i committed:
PR feedback - easier to read `if`

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:
```diff
@@ -196,15 +196,25 @@ class ResourceService(
+      if (list.nonEmpty) throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Cannot delete a resource with children. Delete the children first then try again."))
+    }
+  }
+
```
much cleaner 👍

andy7i replied:
thanks for the feedback!

andy7i committed:
PR feedback - 400 test in ResourceRoutesV2Spec -- needed `  @throws(classOf[WorkbenchExceptionWithErrorReport])
` on the method to make mockito happy about checked exceptions

dvoet committed:
revised action checking on parents

andy7i commented on src/main/resources/swagger/api-docs.yaml:
```diff
@@ -1720,10 +1720,10 @@ paths:
           content: {}
         403:
-          description: You do not have permission to set the parent for the child resource, to remove the child of the current parent, or to add a child to the new parent
+          description: You do not have permission to set the parent for the child resource, to remove the child of the current parent, to add a child to the new parent, or the new parent does not exist
```
After looking at IETF's descriptions of http codes (https://tools.ietf.org/html/rfc7231#section-6.5.1) as well as this SO post (https://stackoverflow.com/questions/48449192/rest-http-response-code-for-missing-parent-resource#:~:text=YES%3A%20A%20404%20Not%20Found,%22sorry%2C%20resource%20unavailable%22.), it seems like there isn't a great option, but a 404 might be the most straightforward option. I could also see a 400 as a general catch-all. 403 feels like a bit of a stretch for "parent resource does not exist", but I'm not super tied to any status code as long as the error message is descriptive and useful for the user.

dvoet replied:
The problem is that we can't let our error conditions reveal the existence of a resource. A user should not be able to tell the difference between no access and does not exist. So, yes, 403 is a stretch for parent resource does not exist but 400 is also a stretch for access denied to the parent and we need to have the same response for both.

andy7i replied:
That makes sense -- we don't want to reveal the existence of a resource. In other places, for a target resource we use 404 to indicate either `no access` or `does not exist`. It feels like we could be consistent by doing that here for parent resources as well. The user would get an error message saying "404, the parent resource does not exist or you have no access."

dvoet replied:
I think that is inconsistent with the HTTP spec. 404 means that some part of the URL does not exist. That is not what is happening here.

andy7i commented on src/main/resources/swagger/api-docs.yaml:
```diff
@@ -1754,7 +1754,7 @@ paths:
           content: {}
         403:
-          description: You do not have permission to set the parent for the child resource or to remove the child of the current parent
+          description: You do not have permission to set the parent for the this resource or to remove this resource of its current parent
```
```suggestion
          description: You do not have permission to set the parent for this resource or to remove this (child) resource from its current parent
```

andy7i commented on src/main/resources/swagger/api-docs.yaml:
```diff
@@ -1720,10 +1720,10 @@ paths:
           content: {}
         404:
-          description: You do not have access to the child, new parent, or current resource, or they do not exist
+          description: You do not have access to this resource, it does not exist
```
```suggestion
          description: You do not have access to this resource, or it does not exist
```

andy7i commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SecurityDirectives.scala:
```diff
@@ -3,48 +3,99 @@ package org.broadinstitute.dsde.workbench.sam.api
+    parentIO.flatMap {
+      case Some(resourceParent) => hasPermissionOneOf(resourceParent, actions, userId, samRequestContext)
+      case None =>
+        // there is no parent so permission is granted
+        IO.pure(true)
```
why do we return true if no parent exists here? It feels weird since we're checking if the user has some permission on the parent. If the parent doesn't exist they wouldn't be able to make changes on the parent right?

dvoet replied:
If there is no parent this effectively disables the parent permission check which seems appropriate.

We hit this case in requiring `remove_child` only at the moment. If there is not parent `remove_child` should be allowed.

andy7i commented on src/test/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutesV2Spec.scala:
```diff
@@ -406,7 +408,7 @@ class ResourceRoutesV2Spec extends FlatSpec with Matchers with TestSupport with
+      status shouldEqual StatusCodes.Forbidden
     }
   }
 
```
I think we're missing a test for erroring when a parent is set on a resource, but the parent resource does not exist

dvoet replied:
can you add what is missing?

andy7i commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SecurityDirectives.scala:
```diff
@@ -3,48 +3,99 @@ package org.broadinstitute.dsde.workbench.sam.api
+        if (hasPermission) {
+          innerRoute
+        } else {
+          val forbiddenErrorMessage = s"You may not perform any of ${parentActions.mkString("[", ", ", "]").toUpperCase} on parent of ${childResource.resourceTypeName.value}/${childResource.resourceId.value}"
```
```suggestion
          val forbiddenErrorMessage = s"You may not perform any of ${parentActions.mkString("[", ", ", "]").toUpperCase} on parent of ${childResource.resourceTypeName.value}/${childResource.resourceId.value}. The parent resource may not exist."
```

If a user tries to add/remove a parent that doesn't exist, they will get this message, right? If that is the case then a more descriptive message like this may be helpful?

dvoet replied:
I put in a different fix

dvoet committed:
error message

dvoet committed:
Apply suggestions from code review

Co-authored-by: Andrew <andy7i@users.noreply.github.com>

andy7i committed:
add unit test for PUT /parent, where it should 404 if the parent resource does not exist

andy7i commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala:
```diff
@@ -339,34 +338,28 @@ trait ResourceRoutes extends UserInfoDirectives with SecurityDirectives with Sam
-            requireAction(resource, SamResourceActions.setParent, userInfo.userId, samRequestContext) {
-              complete(resourceService.deleteResourceParent(resource, samRequestContext).map(_ => StatusCodes.NoContent))
+      requireAction(resource, SamResourceActions.setParent, userInfo.userId, samRequestContext) {
+        requireParentAction(resource, None, SamResourceActions.removeChild, userInfo.userId, samRequestContext) {
```
~I think that `requireParentAction(resource, None` will always pass. Instead of `None` I think we need to to pass in a `currentParent: Option[FullyQualifiedResourceId]` or look for it within `requireParentAction`~

dvoet replied:
see SecurityDirectives line 93: when None is passed it will look up the current parent

andy7i replied:
For `setResourceParent`, if we pass in `Option(newResourceParent)` instead of `None`, the resulting error would be: ~`StatusCodes.BadRequest, "Cannot set parent as this would introduce a cyclical resource hierarchy"`.~

it should get caught as not having the right permissions on the parent and then 403

andy7i replied:
(I missed the parameter name `newParent` in `requireParentAction`)

andy7i commented on src/test/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutesV2Spec.scala:
```diff
@@ -311,6 +323,22 @@ class ResourceRoutesV2Spec extends FlatSpec with Matchers with TestSupport with
+    }
+  }
+
+  it should "404 if the parent resource does not exist" in {
```
This is the new unit test. 

~(It seems like if a user wanted to scrape the names of all existing resources, they could use setResourceParent, which currently gives a 404 when the resource does not exist and possibly a 403 when it does. -- ~Needs more investigation~.)~

dvoet replied:
I don't think this is right. We want a 403 if the parent resource does not exist. It should behave exactly the same if it exists but there is not access. The user should not be able to distinguish between no access to parent or parent does not exist.

andy7i replied:
~looks like ^ is the behavior at the moment, mostly because we pass `None` into `requireParentAction(resource, None` which ends up calling getResourceParent. If we change that, we get the `cyclical resource hierarchy ` error message mentioned in the above comment.~

Nope, nevermind.. seems like it should 403 in either case

andy7i committed:
remove duplicate comment

andy7i committed:
fix unit tests, found an issue with error messaging

andy7i committed:
don't doubt doug

andy7i committed:
Merge remote-tracking branch 'origin/develop' into aj-check-delete-resources-with-children--ca-930

# Conflicts:
#	src/test/scala/org/broadinstitute/dsde/workbench/sam/openam/MockAccessPolicyDAO.scala

andy7i committed:
update mock func return type

andy7i committed:
Mock out parent calls

Since deleting any resource (like groups) requires checking for children now, there are 8 tests across 4 test suites that would need resourceService mocked out. Instead of mocking out all of resourceService for those 4 test suites, it seems like it might be worth living with these mocked functions.

failed tests:
```
[info] *** 8 TESTS FAILED ***
[error] Failed: Total 679, Failed 8, Errors 0, Passed 671, Pending 1
[error] Failed tests:
[error]         org.broadinstitute.dsde.workbench.sam.api.ManagedGroupRoutesV1Spec
[error]         org.broadinstitute.dsde.workbench.sam.api.ResourceRoutesSpec
[error]         org.broadinstitute.dsde.workbench.sam.api.ResourceRoutesV1Spec
[error]         org.broadinstitute.dsde.workbench.sam.api.ManagedGroupRoutesSpec
```

andy7i commented:
jenkins retest -- failure in orch seems unrelated

dvoet commented on src/test/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutesV2Spec.scala:
```diff
@@ -107,20 +107,22 @@ class ResourceRoutesV2Spec extends FlatSpec with Matchers with TestSupport with
+          when(samRoutes.policyEvaluatorService.listResourceAccessPoliciesForUser(mockitoEq(newParent), mockitoEq(defaultUserInfo.userId), any[SamRequestContext]))
+            .thenReturn(IO(Set[AccessPolicyWithoutMembers]()))
+        }
+      case None =>
```
since there is nothing actually in the None case, does this actually change anything? If not, I would undo.

andy7i replied:
yeah that sounds good. When I was reviewing my code, I thought it was a little easier to read this way so I left it as is, but I'm not super attached to it!

andy7i committed:
undo `newParentOpt.map` change
